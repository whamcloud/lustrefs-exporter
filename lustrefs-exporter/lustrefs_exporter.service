[Unit]
Description=Prometheus exporter for Lustre filesystem
Documentation=https://github.com/whamcloud/lustrefs-exporter

[Service]
Environment=RUST_LOG=info
Restart=on-failure
ExecStart=/usr/bin/lustrefs_exporter
MemoryHigh=1750M
MemoryMax=2G

[Install]
WantedBy=multi-user.target
