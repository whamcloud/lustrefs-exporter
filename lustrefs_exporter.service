[Unit]
Description=Prometheus exporter for Lustre filesystem
Documentation=https://github.com/whamcloud/lustrefs-exporter

[Service]
Restart=on-failure
User=prometheus
ExecStart=/usr/bin/lustrefs_exporter

[Install]
WantedBy=multi-user.target

