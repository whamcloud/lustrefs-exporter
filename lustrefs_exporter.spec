Name:           lustrefs_exporter
Version:        0.2.0
Release:        1%{?dist}
Summary:        prometheus exporter for lustre
License:        MIT

Requires(pre): shadow-utils

%description
Prometheus exporter for the Lustre filesystem

%global debug_package %{nil}

%prep

%build
cargo build --release

%install
install -v -d %{buildroot}%{_bindir}
install -v -d %{buildroot}%{_unitdir}
install -v -m 0644 lustrefs_exporter.service %{buildroot}%{_unitdir}
install -v target/release/lustrefs-exporter %{buildroot}%{_bindir}
%{__ln_s} lustrefs-exporter %{buildroot}%{_bindir}/lustrefs_exporter

%files
%{_bindir}/lustrefs-exporter
%{_bindir}/lustrefs_exporter
%{_unitdir}/lustrefs_exporter.service

%pre
getent group prometheus >/dev/null || groupadd -r prometheus
getent passwd prometheus >/dev/null || \
  useradd -r -g prometheus -s /sbin/nologin \
          -c "Prometheus services" prometheus

%post
%systemd_post %{name}.service

%preun
%systemd_preun %{name}.service

%postun
%systemd_postun %{name}.service
