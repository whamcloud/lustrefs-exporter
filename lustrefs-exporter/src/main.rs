// Copyright (c) 2025 DDN. All rights reserved.
// Use of this source code is governed by a MIT-style
// license that can be found in the LICENSE file.

use clap::Parser;
use lustrefs_exporter::{
    Error,
    routes::{
        AppState, app, jobstats_metrics_cmd, lnet_stats_output, lustre_metrics_output,
        net_show_output,
    },
};
use std::{collections::HashMap, net::SocketAddr};

const LUSTREFS_EXPORTER_PORT: &str = "32221";

#[derive(Debug, Parser)]
pub struct CommandOpts {
    /// Port that exporter will listen to
    #[clap(short, long, env = "LUSTREFS_EXPORTER_PORT", default_value = LUSTREFS_EXPORTER_PORT)]
    pub port: u16,

    /// Dump stats as raw string and exit
    #[clap(long, hide = true)]
    dump: bool,
}

/// Dumps Lustre filesystem statistics to stdout
///
/// This function executes several Lustre commands and prints their raw output:
/// - `lctl get_param` with all standard parameters from the parser
/// - `lctl get_param` for jobstats (OST and MDT job statistics)
/// - `lnetctl net show -v 4` for network configuration details
/// - `lnetctl stats show` for network statistics
///
/// # Arguments
/// * `cmd_hdl` - Command handler implementing `RemoteCmd` trait for executing commands
///
/// # Returns
/// * `Ok(())` on successful execution of all commands
/// * `Err(Error)` if any command fails or output cannot be converted to UTF-8
///
/// # Example
/// ```rust
/// use lustrefs_exporter::remote_cmd::LocalCmd;
///
/// dump_stats(&LocalCmd).await?;
/// ```
async fn dump_stats(env_vars: &HashMap<&'static str, String>) -> Result<(), Error> {
    println!("# Dumping lctl get_param output");

    let mut lctl = lustre_metrics_output(env_vars);

    let lctl = lctl.output().await?;

    println!("{}", std::str::from_utf8(&lctl.stdout)?);

    println!("# Dumping lctl get_param jobstats output");

    let mut lctl = jobstats_metrics_cmd(env_vars);

    let lctl = tokio::task::spawn_blocking(move || lctl.output()).await??;

    println!("{}", std::str::from_utf8(&lctl.stdout)?);

    println!("# Dumping lnetctl net show output");

    let mut lnetctl = net_show_output(env_vars);

    let lnetctl = lnetctl.output().await?;

    println!("{}", std::str::from_utf8(&lnetctl.stdout)?);

    println!("# Dumping lnetctl stats show output");

    let mut lnetctl_stats_output = lnet_stats_output(env_vars);

    let lnetctl_stats_output = lnetctl_stats_output.output().await?;

    println!("{}", std::str::from_utf8(&lnetctl_stats_output.stdout)?);

    Ok(())
}

#[tokio::main]
async fn main() -> Result<(), Error> {
    tracing_subscriber::fmt::init();

    let app_state = AppState {
        env_vars: Default::default(),
    };

    let opts = CommandOpts::parse();

    if opts.dump {
        dump_stats(&app_state.env_vars).await?;
    } else {
        let addr = SocketAddr::from(([0, 0, 0, 0], opts.port));

        tracing::info!("Listening on http://{addr}/metrics");

        let listener = tokio::net::TcpListener::bind(("0.0.0.0", opts.port)).await?;

        axum::serve(listener, app(app_state))
            .with_graceful_shutdown(async {
                tokio::signal::ctrl_c().await.ok();
            })
            .await?;
    }

    Ok(())
}

#[cfg(test)]
mod tests {
    use lustrefs_exporter::TestEnv;

    use crate::dump_stats;

    #[tokio::test]
    async fn test_dump_stats() {
        let test_env = TestEnv::default();

        dump_stats(&test_env.vars()).await.unwrap();
    }
}
