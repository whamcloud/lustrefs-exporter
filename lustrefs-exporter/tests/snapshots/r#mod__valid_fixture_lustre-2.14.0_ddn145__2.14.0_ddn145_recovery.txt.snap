---
source: lustrefs-exporter/src/main.rs
expression: x
---

